pkg/go/graph/weighted_graph_builder.go

```diff
@@ -18,60 +21,182 @@ func NewWeightedAuthorizationModelGraphBuilder() *WeightedAuthorizationModelGrap
 	}
 
-	wb := NewWeightedAuthorizationModelGraph()
-	// Add all nodes
-	iterNodes := g.Nodes()
-	for iterNodes.Next() {
-		nextNode := iterNodes.Node()
-		node, ok := nextNode.(*AuthorizationModelNode)
-		if !ok {
-			return nil, fmt.Errorf("%w: could not cast to WeightedAuthorizationModelNode", ErrBuildingGraph)
+	return wb, nil
+}
+
+func (wgb *WeightedAuthorizationModelGraphBuilder) parseRewrite(wg *WeightedAuthorizationModelGraph, parentNode *WeightedAuthorizationModelNode, model *openfgav1.AuthorizationModel, rewrite *openfgav1.Userset, typeDef *openfgav1.TypeDefinition, relation string) error {
+	var operator string
+
+	var children []*openfgav1.Userset
+
+	switch rw := rewrite.GetUserset().(type) {
+	case *openfgav1.Userset_This:
+		err := wgb.parseThis(wg, parentNode, typeDef, relation)
+		return err
+	case *openfgav1.Userset_ComputedUserset:
+		wgb.parseComputed(wg, parentNode, typeDef, rw.ComputedUserset.GetRelation())
+		return nil
+	case *openfgav1.Userset_TupleToUserset:
+		err := wgb.parseTupleToUserset(wg, parentNode, model, typeDef, rw.TupleToUserset)
+		return err
+	case *openfgav1.Userset_Union:
+		operator = UnionOperator
+		children = rw.Union.GetChild()
+
+	case *openfgav1.Userset_Intersection:
+		operator = IntersectionOperator
+		children = rw.Intersection.GetChild()
+
+	case *openfgav1.Userset_Difference:
+		operator = ExclusionOperator
+		children = []*openfgav1.Userset{
+			rw.Difference.GetBase(),
+			rw.Difference.GetSubtract(),
+		}
+	}
+
+	operatorNodeName := fmt.Sprintf("%s:%s", operator, ulid.Make().String())
+	operatorNode := wg.GetOrAddNode(operatorNodeName, operator, OperatorNode)
+
+	// add one edge "relation" -> "operation that defined the operator"
+	// Note: if this is a composition of operators, operationNode will be nil and this edge won't be added.
+	wg.AddEdge(parentNode.GetUniqueLabel(), operatorNodeName, RewriteEdge, "", nil)
+	for _, child := range children {
+		err := wgb.parseRewrite(wg, operatorNode, model, child, typeDef, relation)
+		if err != nil {
+			return err
 		}
+	}
+	return nil
+}
 
-		wb.AddNode(node.uniqueLabel, node.label, node.nodeType)
+func (wgb *WeightedAuthorizationModelGraphBuilder) parseTupleToUserset(wg *WeightedAuthorizationModelGraph, parentNode *WeightedAuthorizationModelNode, model *openfgav1.AuthorizationModel, typeDef *openfgav1.TypeDefinition, rewrite *openfgav1.TupleToUserset) error {
+	// e.g. define viewer: admin from parent
+	// "parent" is the tupleset
+	tuplesetRelation := rewrite.GetTupleset().GetRelation()
+	// "admin" is the computed relation
+	computedRelation := rewrite.GetComputedUserset().GetRelation()
+
+	// find all the directly related types to the tupleset
+	relationMetadata, ok := typeDef.GetMetadata().GetRelations()[tuplesetRelation]
+	if !ok {
+		return fmt.Errorf("%w: Model cannot be parsed. %s invalid tupleset relation", ErrInvalidModel, tuplesetRelation)
 	}
+	directlyRelated := relationMetadata.GetDirectlyRelatedUserTypes()
+
+	validTTU := false
```

Comment:
instead of keep setting validTTU, would it make sense to use the following
```
if len(directlyRelated) == 0 {
    return fmt.Errorf("%w: Model cannot be parsed. No type and relation link exists for tupleset relation %s and computed relation %s", ErrInvalidModel, tuplesetRelation, computedRelation)
}
```